app/controllers/products_controller.rb

```diff
@@ -1,13 +1,29 @@
 class ProductsController < ApplicationController
   before_action :prepare_category_ids, :filter_products, only: :index
+  before_action :load_product_or_redirect, :related_products, only: :show
 
   def index
     @products_found = @products.count
     @products = @products.page(params[:page])
   end
 
+  def show; end
+
   private
 
+  def load_product_or_redirect
+    @product = Product.find_by(id: params[:id])
+    return if @product.present?
```

Comment:
```suggestion
    return if @product
```
ri là đủ rồi, vì với find_by nếu có thì trả về object đầu tiên tìm thấy, nếu ko thì trả về nil rồi